RandomMCEPlayer.java

```diff
@@ -0,0 +1,296 @@
+	// From plays index [numPlays] onward, we maintain a list of yet unplayed positions.
+	private int numPlays = 0; // number of Cards played into the grid so far
+	private PokerSquaresPointSystem system; // point system
+	private int depthLimit = 2; // default depth limit for Random Monte Carlo (MC) play
+	private Card[][] grid = new Card[SIZE][SIZE]; // grid with Card objects or null (for empty positions)
+	private Card[] simDeck = Card.getAllCards(); // a list of all Cards. As we learn the index of cards in the play deck,
+	                                             // we swap each dealt card to its correct index.  Thus, from index numPlays
+												 // onward, we maintain a list of undealt cards for MC simulation.
+	private int[][] legalPlayLists = new int[NUM_POS][NUM_POS]; // stores legal play lists indexed by numPlays (depth)
+	// (This avoids constant allocation/deallocation of such lists during the selections of MC simulations.)
+
+	/**
+	 * Create a Random Monte Carlo player that simulates random play to depth 2.
+	 */
+	public RandomMCPlayer() {
+	}
+
+	/**
+	 * Create a Random Monte Carlo player that simulates random play to a given depth limit.
+	 * @param depthLimit depth limit for random simulated play
+	 */
+	public RandomMCPlayer(int depthLimit) {
+		this.depthLimit = depthLimit;
+	}
+
+	/* (non-Javadoc)
+	 * @see PokerSquaresPlayer#init()
+	 */
+	@Override
+	public void init() {
+		// clear grid
+		for (int row = 0; row < SIZE; row++)
+			for (int col = 0; col < SIZE; col++)
+				grid[row][col] = null;
+		// reset numPlays
+		numPlays = 0;
+		// (re)initialize list of play positions (row-major ordering)
+		for (int i = 0; i < NUM_POS; i++)
+			plays[i] = i;
+	}
+
+	/* (non-Javadoc)
+	 * @see PokerSquaresPlayer#getPlay(Card, long)
+	 */
+	@Override
+	public int[] getPlay(Card card, long millisRemaining) {
+		/*
+		 * With this algorithm, the player chooses the legal play that has the highest expected score outcome.
+		 * This outcome is estimated as follows:
+		 *   For each move, many simulated random plays to the set depthLimit are performed and the (sometimes
+		 *     partially-filled) grid is scored.
+		 *   For each play simulation, random undrawn cards are drawn in simulation and the player
+		 *     picks a play position randomly.
+		 *   After many such plays, the average score per simulated play is computed.  The play with the highest
+		 *     average score is chosen (breaking ties randomly).
+		 */
+
+		// match simDeck to actual play event; in this way, all indices forward from the card contain a list of
+		//   undealt Cards in some permutation.
+		int cardIndex = numPlays;
+		int remainingPlays = NUM_POS - numPlays;
+
+		while (!card.equals(simDeck[cardIndex]))
+			cardIndex++;
+		simDeck[cardIndex] = simDeck[numPlays];
+		simDeck[numPlays] = card;
+
+		// Ben Myles
+		// 2021-11-18
+		if (remainingPlays <= EXPECTISIM_DEPTH) { // Last few turns will be calculated by expectimax.
```

Comment:
We still need to have an extra check to exclude the last play. Maybe something like this:
```suggestion
		if (remainingPlays == 1) { 
			// do nothing (will pick the forced final play automatically)
		}
		else if (remainingPlays <= EXPECTISIM_DEPTH) { // Last few turns will be calculated by expectimax.
```